server/controllers/DatabaseController.js

```diff
@@ -0,0 +1,242 @@
+    );
+  } finally {
+    await client.end();
+    res.write('Disconnected from the database.\n');
+  }
+};
+
+const dumpPostgresDatabase = async (req, res) => {
+  const { dumpOutputDirectory, dumpFileName } = req.body;
+
+  if (dumpFileName && !dumpFileName.includes('.sql'))
+    return res.status(400).send("Provided file name does not include '.sql'");
+
+  const OUTPUT_DIR = dumpOutputDirectory || path.join(os.homedir(), 'db_dumps');
+  const DUMP_FILE = path.join(
+    OUTPUT_DIR,
+    dumpFileName ||
+      `${process.env.ENVIRONMENT}_dump_${dates.convertDateToString(
+        new Date(),
+        'YYYYMMDD_HHmmss'
+      )}.sql`
+  );
+
+  try {
+    // pg_dump command
+    const pgDumpCommand = `PGPASSWORD=${DB_PASSWORD} pg_dump -U ${DB_USER} -h ${DB_HOST} -d ${DB_NAME} --exclude-table=session > ${DUMP_FILE}`;
+
+    // Execute the command
+    exec(pgDumpCommand, (error, stdout, stderr) => {
+      if (error) {
+        return res
+          .status(400)
+          .send(`Error executing pg_dump: ${sanitizeError(error.message)}`);
+      }
+      if (stderr) {
+        return res
+          .status(400)
+          .send(`pg_dump stderr:: ${sanitizeError(stderr)}`);
+      }
+      return res
+        .status(200)
+        .send(`Database dump completed successfully: ${DUMP_FILE}`);
+    });
+  } catch (error) {
+    return res
+      .status(400)
+      .send(`Unable to dump database: ${sanitizeError(error.message)}`);
+  }
+};
+
+const restorePostgresDatabase = async (req, res) => {
+  // Prevent unintentionally dropping or restoring database tables if ran on
+  // production unless manually done
+  if (process.env.ENVIRONMENT === 'production') {
+    return res.status(405).send('This request is not permitted');
+  }
+
+  const { pathToFile } = req.body;
+  if (!pathToFile)
+    return res.status(400).send("'pathToFile' is missing. Please provide.");
+
+  // Purge the database's tables and functions to make restoring with the
+  // pg import easier
+  await removeDatabaseTablesAndFunctions(res);
+
+  try {
+    // delete the tables currently in the db
+    const pgImportCommand = `PGPASSWORD=${DB_PASSWORD} psql -U ${DB_USER} -h ${DB_HOST} -d ${DB_NAME} -f ${pathToFile}`;
+
+    // Execute the command
+    exec(pgImportCommand, (error, stdout, stderr) => {
+      if (error) {
+        res
+          .status(400)
+          .write(`Error executing pg import: ${sanitizeError(error.message)}`);
+      }
+      if (stderr) {
+        res.status(400).write(`pg import stderr:: ${sanitizeError(stderr)}`);
+      }
+      res.status(200).write(`Database import completed successfully`);
+      return res.end();
+    });
+  } catch (error) {
+    res
+      .status(400)
+      .write(`Unable to import database: ${sanitizeError(error.message)}`);
+    return res.end();
+  }
+};
+
+// Function to clean up the dump folder
+const cleanFolder = (req, res) => {
+  const { dumpOutputDirectory, maxFiles } = req.body;
```

Comment:
I think some safety checks are needed here. If I am not mistaken a 
```
{
  "dumpOutputDirectory": "../../../etc",
  "maxFiles": 1
}
``` 
could wreak havoc. I know that user permissions will probably prevent this but with deletion I like to have multiple safeguards.

Maybe this should be exclusively allowed with the default path? Alternatively, some path validation could be used if you want to allow a few different locations. I also think this route shouldn't be accessible on prod either.

Reply:
Good point! Going over this again, I feel it unnecessary to even provide these custom paths for the clean and dump functions